fdbclient/DatabaseBackupAgent.actor.cpp

```diff
@@ -628,31 +662,85 @@ namespace dbBackup {
 		}
 
 		ACTOR static Future<Void> _execute(Database cx, Reference<TaskBucket> taskBucket, Reference<FutureBucket> futureBucket, Reference<Task> task) {
-			state Reference<FlowLock> lock(new FlowLock(CLIENT_KNOBS->BACKUP_LOCK_BYTES));
+			// state Reference<FlowLock> lock(new FlowLock(CLIENT_KNOBS->BACKUP_LOCK_BYTES));
 
 			wait(checkTaskVersion(cx, task, CopyLogRangeTaskFunc::name, CopyLogRangeTaskFunc::version));
 
 			state Version beginVersion = BinaryReader::fromStringRef<Version>(task->params[DatabaseBackupAgent::keyBeginVersion], Unversioned());
 			state Version endVersion = BinaryReader::fromStringRef<Version>(task->params[DatabaseBackupAgent::keyEndVersion], Unversioned());
-			state Version newEndVersion = std::min(endVersion, (((beginVersion-1) / CLIENT_KNOBS->BACKUP_BLOCK_SIZE) + 2 + (g_network->isSimulated() ? CLIENT_KNOBS->BACKUP_SIM_COPY_LOG_RANGES : 0)) * CLIENT_KNOBS->BACKUP_BLOCK_SIZE);
+			
+			Version newEndVersion = std::min(
+			  endVersion, 
+			  (
+			    ((beginVersion-1) / CLIENT_KNOBS->COPY_LOG_BLOCK_SIZE) + 1 
+			    + CLIENT_KNOBS->COPY_LOG_BLOCKS_PER_TASK 
+			    + (g_network->isSimulated() ? CLIENT_KNOBS->BACKUP_SIM_COPY_LOG_RANGES : 0)
+			  ) * CLIENT_KNOBS->COPY_LOG_BLOCK_SIZE
+			);
+
+			state Standalone<VectorRef<KeyRangeRef>> ranges = getLogRanges(beginVersion, newEndVersion, task->params[BackupAgentBase::destUid], CLIENT_KNOBS->COPY_LOG_BLOCK_SIZE);
+			state int nRanges = ranges.size();
+
+			// TraceEvent(SevInfo, "debug.CopyLogRangeTaskFunc::_execute.10").detail("beginVersion", beginVersion).detail("endVersion", endVersion).detail("newEndVersion", newEndVersion).detail("nRanges", nRanges);
 
-			state Standalone<VectorRef<KeyRangeRef>> ranges = getLogRanges(beginVersion, newEndVersion, task->params[BackupAgentBase::destUid], CLIENT_KNOBS->BACKUP_BLOCK_SIZE);
 			state std::vector<PromiseStream<RCGroup>> results;
 			state std::vector<Future<Void>> rc;
-			state std::vector<Future<Void>> dump;
-
-			for (int i = 0; i < ranges.size(); ++i) {
-				results.push_back(PromiseStream<RCGroup>());
-				rc.push_back(readCommitted(taskBucket->src, results[i], Future<Void>(Void()), lock, ranges[i], decodeBKMutationLogKey, true, true, true));
-				dump.push_back(dumpData(cx, task, results[i], lock.getPtr(), taskBucket));
+			state std::vector<Reference<FlowLock>> locks;
+			state Version nextVersion = beginVersion;
+			state double breakTime = timer_monotonic() + CLIENT_KNOBS->COPY_LOG_TASK_DURATION_NANOS;
+			state int rangeN = 0;
+			
+			loop {
+			  if (rangeN >= nRanges)
+			    break;
+
+			  // prefetch
+			  int prefetchTo = std::min(rangeN + CLIENT_KNOBS->COPY_LOG_PREFETCH_BLOCKS, nRanges);
+			  
+			  for (int j = results.size(); j < prefetchTo; j ++) {
+			    results.push_back(PromiseStream<RCGroup>());
+			    locks.push_back(Reference<FlowLock>(new FlowLock(CLIENT_KNOBS->BACKUP_LOCK_BYTES)));
```

Comment:
So you are changing locks from a single one to one for each result. And the memory usage can be increased by a multiple that is controlled by `COPY_LOG_PREFETCH_BLOCKS`. It might be worth to add a release note of your changes and document new knobs.

Reply:
The problem of the single lock is that (n+2) chunk can exhouse the limit before filling (n), so a deadlock will occure.

I agree to document this. Could you give me where may I do it and some example?

Reply:
You can document your changes to `documentation/sphinx/source/release-notes/release-notes-700.rst` file, probably under performance section. See `documentation/sphinx/source/release-notes/release-notes-630.rst` for examples.

Reply:
I added a line to the release notes, but I coudn't find any document with detail description of knobs.

Reply:
I think very large lock is not necessary with prefetch. So I introduced a new knob ``COPY_LOG_READ_AHEAD_BYTES`` with the default value ``BACKUP_LOCK_BYTES`` / ``COPY_LOG_PREFETCH_BLOCKS`` so the memory usage limit is not changed

Reply:
Where can I document the new knobs with details?